streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java

```diff
@@ -173,8 +175,466 @@ public void shouldNameProcessorsAndStoreBasedOnNamedParameter() {
+                        "      <-- KSTREAM-SOURCE-0000000001\n" +
+                        "    Processor: COGROUPKSTREAM-MERGE-0000000010 (stores: [])\n" +
+                        "      --> none\n" +
+                        "      <-- COGROUPKSTREAM-AGGREGATE-0000000008, COGROUPKSTREAM-AGGREGATE-0000000009\n" +
+                        "    Processor: COGROUPKSTREAM-MERGE-0000000017 (stores: [])\n" +
+                        "      --> none\n" +
+                        "      <-- COGROUPKSTREAM-AGGREGATE-0000000015, COGROUPKSTREAM-AGGREGATE-0000000016\n\n"));
+    }
+
+    @Test
+    public void shouldInsertRepartitionsTopicForUpstreamKeyModificationWithGroupedReused() {
+        final StreamsBuilder builder2 = new StreamsBuilder();
+
+        final KStream<String, String> stream1 = builder2.stream("one", stringConsumed);
+        final KStream<String, String> stream2 = builder2.stream("two", stringConsumed);
+
+        final KGroupedStream<String, String> groupedOne = stream1.map((k, v) -> new KeyValue<>(v, k)).groupByKey();
+        final KGroupedStream<String, String> groupedTwo = stream2.groupByKey();
+
+        groupedOne.cogroup(STRING_AGGREGATOR)
+                .cogroup(groupedTwo, STRING_AGGREGATOR)
+                .aggregate(STRING_INITIALIZER);
+
+        groupedOne.aggregate(STRING_INITIALIZER, STRING_AGGREGATOR);
+
+        final String topologyDescription = builder2.build().describe().toString();
+
+        assertThat(
+                topologyDescription,
+                equalTo("Topologies:\n" +
+                        "   Sub-topology: 0\n" +
+                        "    Source: KSTREAM-SOURCE-0000000000 (topics: [one])\n" +
+                        "      --> KSTREAM-MAP-0000000002\n" +
+                        "    Processor: KSTREAM-MAP-0000000002 (stores: [])\n" +
+                        "      --> COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-filter, KSTREAM-FILTER-0000000013\n" +
+                        "      <-- KSTREAM-SOURCE-0000000000\n" +
+                        "    Processor: COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-filter (stores: [])\n" +
+                        "      --> COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-sink\n" +
+                        "      <-- KSTREAM-MAP-0000000002\n" +
+                        "    Processor: KSTREAM-FILTER-0000000013 (stores: [])\n" +
+                        "      --> KSTREAM-SINK-0000000012\n" +
+                        "      <-- KSTREAM-MAP-0000000002\n" +
+                        "    Sink: COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-sink (topic: COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition)\n" +
+                        "      <-- COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-filter\n" +
+                        "    Sink: KSTREAM-SINK-0000000012 (topic: KSTREAM-AGGREGATE-STATE-STORE-0000000010-repartition)\n" +
```

Comment:
Seem there is a naming "gap" -- existing code name the sink `KSTREAM-...` while new code uses `COGROUPKSTREAM` (might be worth to align and use `KSTREAM-...`, in the new code, too?